composer/workflows/s3togcsoperator_tutorial.py

```diff
@@ -0,0 +1,149 @@
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+
+from airflow import models
+from airflow.providers.google.cloud.operators import dataproc
+from airflow.providers.google.cloud.operators.bigquery import BigQueryInsertJobOperator
+from airflow.providers.google.cloud.transfers.gcs_to_bigquery import (
+    GCSToBigQueryOperator,
+)
+from airflow.providers.google.cloud.transfers.s3_to_gcs import S3ToGCSOperator
+from airflow.utils.task_group import TaskGroup
+
+PROJECT_NAME = "{{var.value.gcp_project}}"
+
+# BigQuery configs
+BQ_DESTINATION_DATASET_NAME = "holiday_weather"
+BQ_DESTINATION_TABLE_NAME = "holidays_weather_joined"
+BQ_NORMALIZED_TABLE_NAME = "holidays_weather_normalized"
+
+# Dataproc configs
+BUCKET_NAME = "{{var.value.gcs_bucket}}"
+PYSPARK_JAR = "gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar"
+PROCESSING_PYTHON_FILE = f"gs://{BUCKET_NAME}/data_analytics_process.py"
+
+# S3 configs
+S3_BUCKET_NAME = "{{var.value.s3_bucket}}"
+
+BATCH_ID = "data-processing-{{ ts_nodash | lower}}"  # Dataproc serverless only allows lowercase characters
+BATCH_CONFIG = {
+    "pyspark_batch": {
+        "jar_file_uris": [PYSPARK_JAR],
+        "main_python_file_uri": PROCESSING_PYTHON_FILE,
+        "args": [
+            BUCKET_NAME,
+            f"{BQ_DESTINATION_DATASET_NAME}.{BQ_DESTINATION_TABLE_NAME}",
+            f"{BQ_DESTINATION_DATASET_NAME}.{BQ_NORMALIZED_TABLE_NAME}",
+        ],
+
+    },
+    "environment_config": {
+        "execution_config": {
+            "service_account": "{{var.value.dataproc_service_account}}"
+        }
+    }
+}
+
+yesterday = datetime.datetime.combine(
+    datetime.datetime.today() - datetime.timedelta(1), datetime.datetime.min.time()
+)
+
+default_dag_args = {
+    # Setting start date as yesterday starts the DAG immediately when it is
+    # detected in the Cloud Storage bucket.
+    "start_date": yesterday,
+    # To email on failure or retry set 'email' arg to your email and enable
+    # emailing here.
+    "email_on_failure": False,
+    "email_on_retry": False
+}
+
+with models.DAG(
+    "s3_to_gcs_dag",
+    # Continue to run DAG once per day
+    schedule_interval=datetime.timedelta(days=1),
+    default_args=default_dag_args,
+) as dag:
+
+    s3_to_gcs_op = S3ToGCSOperator(
+        task_id="s3_to_gcs",
+        bucket=S3_BUCKET_NAME,
+        gcp_conn_id="google_cloud_default",
+        aws_conn_id="aws_s3_connection",
+        dest_gcs=f"gs://{BUCKET_NAME}",
+    )
+
+    create_batch = dataproc.DataprocCreateBatchOperator(
+        task_id="create_batch",
+        project_id=PROJECT_NAME,
+        region="{{ var.value.gce_region }}",
```

Comment:
We're using a lot of mix of constants, predefined values and dynamic values. A mix of constants (i.e. `task_id="create_batch"`) and predefined values (i.e. `project_id=PROJECT_NAME`) seem good, we should stick to this style being consistent and also define values like `region="{{ var.value.gce_region }}"`.

Reply:
This is pretty common in Airflow to have predefined values and dynamic values. Dynamic values (which are populated with the jinja templates) can be shared across DAGs. I tend to use constants with dynamic values that are repeated over and over within a single Dag file (like the project_id one). Can you clarify what you mean about defining the region? If it's specifying a region to be prescriptive, I disagree, because this Airflow variable is used in multiple DAG files in the tutorial, so it's best Airflow practice to not have it be static here. If you mean something else though lmk! I am open to it being turned into a constant just to make it look nicer (and if Airflow considers it valid) if that is an okay compromise.

Reply:
Ah I meant initializing a constant for `{{ var.value.gce_region }}`, like

```
PROJECT_NAME = "{{var.value.gcp_project}}"
REGION = "{{ var.value.gce_region }}"

...

    create_batch = dataproc.DataprocCreateBatchOperator(
        task_id="create_batch",
        project_id=PROJECT_NAME,
        region=REGION,
```

Reply:
Done!